doc/dev/backend-plugins.md

```diff
@@ -0,0 +1,271 @@
+In this phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `storage` phase of this `get` operation.
+  The key name and value of this key are read-only.
+- The exact `ks` that was returned by the `storage` phase of this `get` operation.
+
+Again there are no restrictions on what the backend plugin may do in this phase.
+However, unlike the `prestorage` phase, this phase is a very important one.
+It is where validation, generation of implicit values and similar tasks happen.
+
+Finally, `libelektra-kdb` merges the keyset returned by the `poststorage` phase with the ones returned by other backend plugins for different mountpoints and then returns it to the user.
+
+TODO: Are modifications to `parentKey` visible to the user?
+
+### Operation `set`
+
+The `set` operation is optional.
+A mountpoint is automatically read-only and doesn't support the `set` operation, if the backend plugin does not define a `elektra<Plugin>Set` function.
+
+Alternatively, the read-only nature of the mountpoint may also be indicated by the backend plugin during the `init` phase of the `get` operation.
+
+#### Resolver Phase
+
+During the `resolver` phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `resolver` phase of the last `get` or `set` operation.
+  The key name of this key is read-only.
+- An empty keyset `ks`.
+
+The backend plugin then:
+
+- **MUST** set the value of the `parentKey` to a value identifying the storage unit that contains the data of the mountpoint.
+  For file-based backend plugins, this means setting the value of `parentKey` to an absolute filename.
+- **MAY** set metadata on `parentKey`, if encoding the information required for the following phases is too hard to encode in a single string value.
+- **MUST** check whether the data was changed since the last `get` operation.
+  The result of this check is given to `libelektra-kdb` via the return value of the `get` function.
+- **MUST** ensure that errors in the storage phases can be safely rolled back and that the `set` operation does not affect concurrent operations.
+  For file-based backends, this means creating a temporary storage file and returning its absolute filename instead of the name of the actual storage file.
+
+> **Note**: The backend plugin may also modify the keyset `ks`, but `libelektra-kdb` will discard this keyset after this phase, so these modifications won't have any effects.
+
+#### Storage Phases
+
+These phases are responsible for validating and writing data to the KDB.
+
+In the `prestorage` phase the backend plugin is called with:
+
+- The exact `parentKey` that was returned by the `resolver` phase of this `set` operation.
+  The key name and value of this key are read-only.
+- The exact subset below `parentKey` of the keyset `ks` that was provided by the user.
+
+There are no restrictions on what the backend plugin may do in this phase.
```

Comment:
Some parts are very repetitive. Is it possible to put this to some intro text?

Reply:
The repetition is intentional, I tried to keep a consistent schema so the differences between the phases are obvious. I don't think an intro would help, but probably some formatting and maybe using table would be a good idea.

Reply:
Yes, improving the formatting/headings or some other indication that now a repetitive text comes would be very helpful.